File: packages/examples/todo-app/src/commands/stats.ts
/**
 * @fileoverview todo.stats command
 */

import { z } from 'zod';
import { defineCommand, success } from '@afd/server';
import { store } from '../store/memory.js';
import type { TodoStats } from '../types.js';

const inputSchema = z.object({});

export const getStats = defineCommand<typeof inputSchema, TodoStats>({
	name: 'todo.stats',
	description: 'Get todo statistics',
	category: 'todo',
	mutation: false,
	version: '1.0.0',
	input: inputSchema,

	async handler() {
		const stats = store.getStats();

		// Build reasoning with summary
		const parts: string[] = [];

		if (stats.total === 0) {
			parts.push('No todos yet');
		} else {
			parts.push(`${stats.total} total todos`);
			parts.push(`${stats.completed} completed`);
			parts.push(`${stats.pending} pending`);
			parts.push(`${Math.round(stats.completionRate * 100)}% completion rate`);
		}

		return success(stats, {
			reasoning: parts.join(', '),
			confidence: 1.0,
		});
	},
});

File: packages/examples/todo/backends/rust/src/commands/stats.rs
use super::{ok, Command};
use crate::store::TodoStore;
use afd::CommandResult;
use serde::Deserialize;
use serde_json::{json, Value};
use std::sync::Arc;

pub struct Stats(pub Arc<TodoStore>);

#[derive(Deserialize)]
pub struct Input {}

impl Command for Stats {
    type Input = Input;
    const NAME: &'static str = "todo-stats";
    const DESCRIPTION: &'static str = "Get todo statistics";
    const MUTATION: bool = false;

    fn schema() -> Value {
        json!({ "type": "object", "properties": {} })
    }

    fn run(&self, _input: Input) -> CommandResult<Value> {
        let stats = self.0.stats();
        let reasoning = if stats.total == 0 {
            "No todos yet".to_string()
        } else {
            format!(
                "{} total todos, {} completed, {} pending, {}% completion rate",
                stats.total,
                stats.completed,
                stats.pending,
                (stats.completion_rate * 100.0).round()
            )
        };
        ok(&stats, reasoning, 1.0, Vec::new())
    }
}
